lib/imagec/download.go

```diff
@@ -0,0 +1,330 @@
+	return func(progressChan chan<- progress.Progress, start <-chan struct{}, inactive chan<- struct{}) xfer.Transfer {
+
+		d := &downloadTransfer{
+			Transfer: xfer.NewTransfer(),
+			layer:    layer,
+		}
+
+		go func() {
+
+			defer func() {
+				close(progressChan)
+			}()
+
+			progressOutput := progress.ChanOutput(progressChan)
+
+			// wait for TransferManager to give the go-ahead
+			select {
+			case <-start:
+			default:
+				progress.Update(progressOutput, layer.String(), "Waiting")
+				<-start
+			}
+
+			if parentDownload != nil {
+				// bail if parent download failed or was cancelled
+				select {
+				case <-parentDownload.Done():
+					err := parentDownload.result()
+					if err != nil {
+						d.err = err
+						return
+					}
+				default:
+				}
+			}
+
+			// fetch blob
+			diffID, err := FetchImageBlob(d.Transfer.Context(), ic, layer, progressOutput)
+			if err != nil {
+				d.err = fmt.Errorf("%s/%s returned %s", ic.Image, layer.ID, err)
+				return
+			}
+
+			layer.diffID = diffID
+
+			close(inactive)
+
+			if parentDownload != nil {
+				select {
+				case <-d.Transfer.Context().Done():
+					d.err = errors.New("layer download cancelled")
+					return
+				case <-parentDownload.Done(): // block until parent download completes
+				}
+
+				err := parentDownload.result()
+				if err != nil {
+					d.err = err
+					return
+				}
+			}
+
+			// is this the child-most leaf layer?
+			imageLayer := layer.ID == layers[0].ID
+
+			// if this is the child-most layer, we are done and can now create the image config
+			if imageLayer {
+				imageConfig, err := ic.CreateImageConfig(layers)
+				if err != nil {
+					d.err = err
+					return
+				}
+
+				// cache the image
+				cache.ImageCache().AddImage(&imageConfig)
+			}
+
+			ldm.m.Lock()
+			defer ldm.m.Unlock()
+
+			// Write blob to the storage layer
+			if err := ic.WriteImageBlob(layer, progressOutput, imageLayer); err != nil {
+				log.Infof("Got error writing image blob: %s", err)
+				d.err = err
+				return
+			}
+
+			// update tags
+			// TODO(jzt): this does not appear to work any longer
+			if err := updateImageMetadata(ic, layer, ic.ImageManifest); err != nil {
+				d.err = err
+				return
+			}
+
+			ldm.unregisterDownload(layer.ID)
+
+		}()
+
+		return d
+	}
+}
+
+// makeDownloadFuncFromDownload returns a func used by the TransferManager to
+// handle a layer that was already seen in this image pull, or is currently being downloaded
+func (ldm *LayerDownloader) makeDownloadFuncFromDownload(layer *ImageWithMeta, sourceDownload, parentDownload *downloadTransfer, layers []*ImageWithMeta) xfer.DoFunc {
+
+	return func(progressChan chan<- progress.Progress, start <-chan struct{}, inactive chan<- struct{}) xfer.Transfer {
+
+		d := &downloadTransfer{
+			Transfer: xfer.NewTransfer(),
+			layer:    layer,
+		}
+
+		go func() {
+			defer func() {
+				close(progressChan)
```

Comment:
same as above, am I missing something here. why we are not calling it like this?